rts/Game/UI/IKeys.cpp

```diff
@@ -0,0 +1,132 @@
+	return iter->second;
+}
+
+std::string IKeys::GetDefaultName(int code) const
+{
+	const auto pred = [](const CodeNamePair& a, const CodeNamePair& b) { return (a.first < b.first); };
+	const auto iter = std::lower_bound(defaultCodeToName.begin(), defaultCodeToName.end(), CodeNamePair{0, ""}, pred);
+
+	if (iter == defaultCodeToName.end() || iter->first != code)
+		return IntToString(code, "0x%03X");
+
+	return iter->second;
+}
+
+
+bool IKeys::AddKeySymbol(const std::string& name, int code)
+{
+	if ((code < 0) || !IsValidLabel(name))
+		return false;
+
+	const std::string keysym = StringToLower(name);
+
+	// do not allow existing keysyms to be renamed
+	const auto namePred = [](const NameCodePair& a, const NameCodePair& b) { return (a.first < b.first); };
+	const auto nameIter = std::lower_bound(nameToCode.begin(), nameToCode.end(), NameCodePair{keysym, 0}, namePred);
+
+	if (nameIter != nameToCode.end() && nameIter->first == name)
+		return false;
+
+	nameToCode.emplace_back(keysym, code);
+	// assumes that the user would rather see their own names
+	codeToName.emplace_back(code, keysym);
+
+	// swap into position
+	for (size_t i = nameToCode.size() - 1; i > 0; i--) {
+		if (nameToCode[i - 1].first < nameToCode[i].first)
+			break;
+
+		std::swap(nameToCode[i - 1], nameToCode[i]);
+	}
+	for (size_t i = codeToName.size() - 1; i > 0; i--) {
+		if (codeToName[i - 1].first < codeToName[i].first)
+			break;
+
+		std::swap(codeToName[i - 1], codeToName[i]);
+	}
+
+	return true;
+}
+
+
+bool IKeys::IsPrintable(int code) const
+{
+	const auto pred = [](int a, int b) { return (a < b); };
+	const auto iter = std::lower_bound(printableCodes.begin(), printableCodes.end(), code, pred);
+
+	return (iter != printableCodes.end() && *iter == code);
+}
+
+void IKeys::SaveUserKeySymbols(FILE* file) const
+{
+	bool output = false;
+
+	for (const auto& p: nameToCode) {
+		const auto defSymPred = [](const NameCodePair& a, const NameCodePair& b) { return (a.first < b.first); };
+		const auto defSymIter = std::lower_bound(defaultNameToCode.begin(), defaultNameToCode.end(), NameCodePair{p.first, 0}, defSymPred);
+
+		if (defSymIter != defaultNameToCode.end() && defSymIter->first == p.first)
+			continue;
+
+		// this keysym is not standard
+		const std::string& extSymName = p.first;
+		const std::string& defSymName = GetDefaultName(p.second);
+		fprintf(file, "keysym  %-10s  %s\n", extSymName.c_str(), defSymName.c_str());
+		output = true;
+	}
+
+	if (!output)
+		return;
+
+	fprintf(file, "\n");
+}
+
+bool IKeys::IsValidLabel(const std::string& label)
+{
+	if (label.empty())
+		return false;
+
+	if (!isalpha(label[0]))
+		return false;
+
+	// if any character is not alpha-numeric *and* not space, reject label as invalid
+	return (std::find_if(label.begin(), label.end(), [](char c) { return (!isalnum(c) && (c != '_')); }) == label.end());
+}
+
+
+void IKeys::AddPair(const std::string& name, const int code, const bool printable)
+{
+	nameToCode.emplace_back(name, code);
```

Comment:
Not sure if this is a problem but I'm going to mention it. This file uses `lower_bound` over and over, and since it does a binary search it assumes that the containers are sorted. However here you just append new elements to them, and I don't see a sort being done. Is this ok?

Reply:
I also found the name<->code lookup code weird, order does not have any relevancy here. It's just used as a store for int<->keysetalias

Reply:
if order is not relevant, then all the lower_bounds are broken. They do binary search, but if the underlying thing is not sorted, they will find/not find the searched for element completely randomly.

Reply:
Order is not relevant (for the functionality) but it is sorted during emplacement, perhaps just to use this find method: https://github.com/badosu/spring/blob/a4abeaf6e5b9df719f367ea0066603ed26ea06aa/rts/Game/UI/KeyCodes.cpp#L153-L155

Reply:
I decided to not change this for now